lib/src/folding/foldable_block.dart

```diff
@@ -55,38 +55,102 @@ class FoldableBlock extends InclusiveRange with EquatableMixin {
 
     return false;
   }
+
+  bool isSameLines(FoldableBlock other) {
+    return firstLine == other.firstLine && lastLine == other.lastLine;
+  }
+
+  FoldableBlock join(FoldableBlock other) {
+    return FoldableBlock(
+      firstLine: min(firstLine, other.firstLine),
+      lastLine: max(lastLine, other.lastLine),
+      type: _getJoinType(this, other),
+    );
+  }
+
+  static FoldableBlockType _getJoinType(FoldableBlock a, FoldableBlock b) {
+    if (a.type == FoldableBlockType.imports) {
+      return FoldableBlockType.imports;
+    }
+
+    if (b.type == FoldableBlockType.imports) {
+      return FoldableBlockType.imports;
+    }
+
+    return FoldableBlockType.union;
+  }
 }
 
 extension FoldableBlockList on List<FoldableBlock> {
   void sortByStartLine() {
     sort((a, b) => a.firstLine - b.firstLine);
   }
 
-  /// Joins intersecting blocks in list.
-  /// Expected that list is sorted by start line.
+  /// Joins intersecting blocks.
+  ///
+  /// This list must be sorted by firstLine.
   void joinIntersecting() {
     if (length < 2) {
       return;
     }
 
-    for (int i = 1; i < length; i++) {
-      final currentBlock = this[i];
-      final previousBlock = this[i - 1];
-
-      final areIntersected = previousBlock.lastLine >= currentBlock.firstLine &&
-          previousBlock.lastLine < currentBlock.lastLine;
-
-      final isDuplicate = previousBlock.firstLine == currentBlock.firstLine &&
-          previousBlock.lastLine == currentBlock.lastLine;
-
-      if (areIntersected || isDuplicate) {
-        this[i - 1] = FoldableBlock(
-          firstLine: previousBlock.firstLine,
-          lastLine: currentBlock.lastLine,
-          type: FoldableBlockType.union,
-        );
-        removeAt(i);
-        i = max(i - 2, 0);
+    // A working list to lay down blocks to compare.
+    // As we iterate blocks, it will contain the current nesting hierarchy.
+    final ancestors = <FoldableBlock>[];
+    final ancestorIndexToOverallIndex = <int>[];
+
+    for (int overallIndex = 0; overallIndex < length; overallIndex++) {
+      FoldableBlock bubble = this[overallIndex];
+
+      // Throw the block into the current nesting hierarchy...
+      ancestors.add(bubble);
+      ancestorIndexToOverallIndex.add(overallIndex);
+
+      int bubbleIndexInAncestors = ancestors.length - 1;
+      int bubbleOverallIndex = overallIndex;
+
+      // And fix every violation of the hierarchy by bubbling the block up,
+      // removing non-ancestors from the working list, and joining when needed.
+      for (int ancestorIndex = ancestors.length - 1; --ancestorIndex >= 0;) {
```

Comment:
I think this way it is more readable.
```dart
for (int ancestorIndex = ancestors.length - 2; ancestorIndex >= 0; --ancestorIndex)
```
Also emphasizes on why the first iteration should not consider last element of the list.

Reply:
`+`